src/api/query.rs

```diff
@@ -0,0 +1,402 @@
+use actix_web::{get, web, HttpResponse};
+use futures::TryStreamExt;
+use mongodb::{
+    bson::{doc, Document},
+    Client, Collection, IndexModel,
+};
+use std::collections::HashMap;
+
+use crate::{
+    api::util,
+    models::{query_models::*, response},
+};
+
+const DB_NAME: &str = "boom";
+
+// builds find options for mongo query
+pub fn build_options(
+    projection: Option<mongodb::bson::Document>,
+    kwargs: QueryKwargs,
+) -> mongodb::options::FindOptions {
+    let mut find_options = mongodb::options::FindOptions::default();
+    if kwargs.limit.is_some() {
+        find_options.limit = Some(kwargs.limit.unwrap());
+    }
+    if kwargs.skip.is_some() {
+        find_options.skip = Some(kwargs.skip.unwrap());
+    }
+    if kwargs.sort.is_some() {
+        find_options.sort = Some(kwargs.sort.unwrap());
+    }
+    if kwargs.max_time_ms.is_some() {
+        find_options.max_time = Some(std::time::Duration::from_millis(
+            kwargs.max_time_ms.unwrap(),
+        ));
+    }
+    if projection.is_some() {
+        find_options.projection = Some(projection.unwrap());
+    }
+
+    find_options
+}
+
+pub fn build_cone_search_filter(
+    mut filter: mongodb::bson::Document,
+    radec: (f64, f64),
+    mut radius: f64,
+    unit: Unit,
+) -> mongodb::bson::Document {
+    let ra = radec.0 - 180.0;
+    let dec = radec.1;
+    // convert radius to radians based on unit
+    match unit {
+        Unit::Degrees => radius = radius.to_radians(),
+        Unit::Arcseconds => radius = radius.to_radians() / 3600.0,
+        Unit::Arcminutes => radius = radius.to_radians() / 60.0,
+        Unit::Radians => {}
+    }
+    let center_sphere = doc! {
+        "$centerSphere": [[ra, dec], radius]
+    };
+    let geo_within = doc! {
+        "$geoWithin": center_sphere
+    };
+    filter.insert("coordinates.radec_geojson", geo_within);
+    filter
+}
+
+pub async fn get_catalog_names(
+    client: web::Data<Client>,
+    db_name: &str,
+) -> Result<Vec<String>, mongodb::error::Error> {
+    // get collection names in alphabetical order
+    let collection_names = match client.database(db_name).list_collection_names().await {
+        Ok(c) => c,
+        Err(e) => return Err(e),
+    };
+    let mut data = collection_names
+        .iter()
+        .filter(|name| !name.starts_with("system."))
+        .cloned()
+        .collect::<Vec<String>>();
+    data.sort();
+    return Ok(data);
+}
+
+pub async fn get_catalog_info(
+    client: web::Data<Client>,
+    catalogs: Vec<String>,
+    db_name: &str,
+) -> Result<Vec<Document>, mongodb::error::Error> {
+    let mut data = Vec::new();
+    for catalog in catalogs {
+        match client
+            .database(db_name)
+            .run_command(doc! {
+                "collstats": catalog
+            })
+            .await
+        {
+            Ok(d) => data.push(d),
+            Err(e) => return Err(e),
+        };
+    }
+    return Ok(data);
+}
+
+pub async fn get_index_info(
```

Comment:
same comment as in a couple of other places. These methods should just take the db object as a parameter rather than the client + the db name to recreate the db object everytime.